CompressionEngine.java

```diff
@@ -0,0 +1,483 @@
+              context_data_position = 0;
+              context_data_string += (char) context_data_val;
+              context_data_val = 0;
+            } else {
+              context_data_position++;
+            }
+            value = value >> 1;
+          }
+        } else {
+          value = 1;
+          for (int i = 0; i < context_numBits; i++) {
+            context_data_val = (context_data_val << 1) | value;
+            if (context_data_position == 15) {
+              context_data_position = 0;
+              context_data_string += (char) context_data_val;
+              context_data_val = 0;
+            } else {
+              context_data_position++;
+            }
+            value = 0;
+          }
+          value = (int) context_w.charAt(0);
+          for (int i = 0; i < 16; i++) {
+            context_data_val = (context_data_val << 1)
+                | (value & 1);
+            if (context_data_position == 15) {
+              context_data_position = 0;
+              context_data_string += (char) context_data_val;
+              context_data_val = 0;
+            } else {
+              context_data_position++;
+            }
+            value = value >> 1;
+          }
+        }
+        context_enlargeIn--;
+        if (Double.valueOf(context_enlargeIn).intValue() == 0) {
+          context_enlargeIn = Math.pow(2, context_numBits);
+          context_numBits++;
+        }
+        context_dictionaryToCreate.remove(context_w);
+      } else {
+        value = context_dictionary.get(context_w);
+        for (int i = 0; i < context_numBits; i++) {
+          context_data_val = (context_data_val << 1) | (value & 1);
+          if (context_data_position == 15) {
+            context_data_position = 0;
+            context_data_string += (char) context_data_val;
+            context_data_val = 0;
+          } else {
+            context_data_position++;
+          }
+          value = value >> 1;
+        }
+
+      }
+      context_enlargeIn--;
+      if (Double.valueOf(context_enlargeIn).intValue() == 0) {
+        context_enlargeIn = Math.pow(2, context_numBits);
+        context_numBits++;
+      }
+    }
+
+    // Mark the end of the stream
+    value = 2;
+    for (int i = 0; i < context_numBits; i++) {
+      context_data_val = (context_data_val << 1) | (value & 1);
+      if (context_data_position == 15) {
+        context_data_position = 0;
+        context_data_string += (char) context_data_val;
+        context_data_val = 0;
+      } else {
+        context_data_position++;
+      }
+      value = value >> 1;
+    }
+
+    // Flush the last char
+    while (true) {
+      context_data_val = (context_data_val << 1);
+      if (context_data_position == 15) {
+        context_data_string += (char) context_data_val;
+        break;
+      } else
+        context_data_position++;
+    }
+    return context_data_string;
+  }
+
+  public static String decompress(String compressed) {
```

Comment:
Avoid static methods if possible. Instead declared classes like Compressor and Decompressor with methods you can call.

You could even make them interfaces and just use the interfaces everywhere, and instantiate them with the real implementation just in the beginning.